### PR TITLE
Extended environment variable control

### PR DESCRIPTION
**Pull request by richtja**
This PR adds more control over the environment variables in exec-tests for users. It brings a way how to disable one env variable or clear the whole environment during the test runtime so it won't be available to test script.

### CODE REVIEW

**Comment by richtja** [comment]
Hi @dgibson, please have a look and let me know if this satisfies pasta needs from #5901. You can look at [documentation](https://avocado-framework--5901.org.readthedocs.build/en/5901/guides/writer/chapters/basics.html#user-s-environment-variables) which describes environment variable control.

File: docs/source/guides/writer/chapters/basics.rst [hunk 1]
```diff
@@ -11,3 +12,60 @@ Notice that the file is given executable permissions, which is a
 requirement for Avocado to treat it as a executable test. Also notice
 that the script exits with status code 0, which signals a successful
 result to Avocado.
+
+SIMPLE Tests BASH extensions
```
Comment (clebergnu) [inline docs/source/guides/writer/chapters/basics.rst:16]: This may be a good time to rename this section.  We tried to eliminate the ```SIMPLE``` nomenclature with the legacy runner.

File: docs/source/guides/writer/chapters/basics.rst [hunk 2]
```diff
@@ -69,3 +72,21 @@ currently exports to exec-test tests in default:
 +-----------------------------+---------------------------------------+-----------------------------------------------------------------------------------------------------+
 | `***`                       | All variables from --mux-yaml         | TIMEOUT=60; IO_WORKERS=10; VM_BYTES=512M; ...                                                       |
 +-----------------------------+---------------------------------------+-----------------------------------------------------------------------------------------------------+
+
+User's environment variables
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You can also let avocado set your own environment variables. For that, you need
+to pass the environment variables ass kwargs for the test. Here is an example
```
Comment (clebergnu) [inline docs/source/guides/writer/chapters/basics.rst:80]: s/ass/as/

Also, the info about kwargs is not universal, it's only about exec-test (and similars like tap).  So, I'd write:

"pass the environment variables as keyword arguments (``kwargs``) to the ``exec-test``s (and derived like ``tap``)".

**Review by clebergnu — CHANGES_REQUESTED** [review]
Hi @richtja,

Thanks for working on this.  I've left my feedback in individual comments.

Thanks!

File: avocado/plugins/runners/exec_test.py [hunk 3]
```diff
@@ -122,6 +122,9 @@ def _get_env(self, runnable):
         env = dict(os.environ)
         if runnable.kwargs:
             env.update(runnable.kwargs)
+            for key, value in runnable.kwargs.items():
+                if value == "NULL":
```
[the remark this answers is not in the document]
Comment (richtja) [reply avocado/plugins/runners/exec_test.py:130]: Ok I will update that.

File: avocado/plugins/runners/exec_test.py [hunk 4]
```diff
@@ -119,7 +119,11 @@ def _is_uri_a_file_on_cwd(uri):
         return False
 
     def _get_env(self, runnable):
-        env = dict(os.environ)
+        clear_env = runnable.kwargs.get("AVOCADO_CLEAR_ENV", "FALSE") == "TRUE"
```
[the remark this answers is not in the document]
Comment (richtja) [reply avocado/plugins/runners/exec_test.py:122]: Yes, I wasn't sure with that, but I didn't want to clutter up avocado's config with attributes relevant only to one runner. But when I went again the config we already have `runner.exectest.exitcodes.skip` which is an example of the same use-case. Therefore, I will change it to config value.

**Comment by richtja** [comment]
Hi @clebergnu I have applied changes based on your proposals. Please have a look.

File: avocado/plugins/runners/exec_test.py [hunk 5]
```diff
@@ -122,6 +122,9 @@ def _get_env(self, runnable):
         env = dict(os.environ)
         if runnable.kwargs:
             env.update(runnable.kwargs)
+            for key, value in runnable.kwargs.items():
+                if value is None:
```
Comment (clebergnu) [inline avocado/plugins/runners/exec_test.py:134]: I'd be more explicit here and write this as:

```python
for key, value in runnable.kwargs.items():
   if value is None and key in env:
      del(env[key])
```

The reason being there's no need to return any value (which `pop()` does) and the `None` value as a safeguard makes reading the goal of the `pop()` confusing.

File: docs/source/guides/writer/chapters/basics.rst [hunk 6]
```diff
@@ -90,3 +90,12 @@ variable:
 
 .. note:: All environment variables set by avocado will be accessible only during
           the test runtime and it won't change your environment.
+
+Disabling environment variables
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Let's imagine that your testing environment has some important variables, but they
+could have a negative impact on one of your tests. In that case, avocado let you
```
Comment (clebergnu) [inline docs/source/guides/writer/chapters/basics.rst:97]: `s/avocado let you/avocado lets you/`

File: docs/source/guides/writer/chapters/basics.rst [hunk 7]
```diff
@@ -90,3 +90,12 @@ variable:
...
+disable those variables during the test runtime. To disable the test variable,
```
Comment (clebergnu) [inline docs/source/guides/writer/chapters/basics.rst:98]: `s/to disable the test variable/to disable a test variable/`

File: selftests/functional/basic.py [hunk 8]
```diff
@@ -1093,6 +1101,17 @@ def test_non_absolute_path(self):
             f"Avocado did not return rc {expected_rc}:\n{result}",
         )
 
+    @skipUnlessPathExists("/bin/sh")
```
Comment (clebergnu) [inline selftests/functional/basic.py:1104]: I'm assuming `/bin/sh` here is supposed to match `EXEC_ENV_VARIABLE_TEST`, but `/bin/sh` and `/bin/bash` are not the same on every platform.

File: docs/source/guides/writer/chapters/basics.rst [hunk 9]
```diff
@@ -99,3 +99,10 @@ disable those variables during the test runtime. To disable the test variable,
 you need to set it in test `kwargs` to `None`  like this::
 
   Runnable("exec-test", "examples/tests/sleeptest.sh", SLEEP_LENGTH=None)
+
+If you need to clear the whole environment before your test, then you can set
+`runner.exectest.clear_env` config variable to `True`. When this variable is set, avocado will clear
```
Comment (clebergnu) [inline docs/source/guides/writer/chapters/basics.rst:104]: Maybe you meant to use double backticks here too?

File: avocado/core/__init__.py [hunk 10]
```diff
@@ -178,6 +178,15 @@ def register_core_options():
         help_msg=help_msg,
     )
 
+    help_msg = "When this is enabled, avocado will disable all environment variables during exec testing"
```
Comment (clebergnu) [inline avocado/core/__init__.py:181]: The commit message does not matches the changes in this version.  Also maybe simply stating:

Whether to run the exec-test with an empty environment (no variables set)

**Review by clebergnu — CHANGES_REQUESTED** [review]
Hi @richtja,

Thanks for this second version!  I've spotted a few extra things though.

**Comment by richtja** [comment]
Hi @clebergnu thank you very much for your review, It was really helpful. I addressed your comments in `force-push`. Please have a look.

**Comment by dgibson** [comment]
> Hi @dgibson, please have a look and let me know if this satisfies pasta needs from #5901. You can look at [documentation](https://avocado-framework--5901.org.readthedocs.build/en/5901/guides/writer/chapters/basics.html#user-s-environment-variables) which describes environment variable control.

So far, we don't really have much in the way of requirements here.  I suggested something like this to Cleber, because I think long term it's something we'll want, and makes for a more generally useful way of running these "extended exec" tests from Avocado.  I'll review as best I can.

File: selftests/functional/basic.py [hunk 11]
```diff
@@ -98,6 +98,14 @@ def test(self):
         logging.getLogger("some.other.logger").info("SHOULD BE ON debug.log")
 """
 
+EXEC_ENV_VARIABLE_TEST = """#!/bin/bash
+if [[ -z "${TEST_ENV}"  ]]; then
+    exit 1
```
Comment (dgibson) [inline selftests/functional/basic.py:103]: So, this script "fails" when the variable is empty, although that's actually the "good" case below.  IIUC, that's handled by expecting a return code of 1 below, but it still seems a little counterintuitive to me to do this rather than having the script fail if the variable is set.

File: selftests/functional/basic.py [hunk 12]
```diff
@@ -1093,6 +1101,17 @@ def test_non_absolute_path(self):
...
+    @skipUnlessPathExists("/bin/bash")
+    def test_env_var_disable(self):
+        with script.TemporaryScript("exec_env_var.sh", EXEC_ENV_VARIABLE_TEST) as tst:
+            res = process.run(
+                f"env TEST_ENV=test avocado-runner-exec-test runnable-run -k exec-test -u {tst} TEST_ENV=json:null",
+            )
+            result = res.stdout_text.split("\n")[-2]
+            self.assertIn(
+                "'returncode': 1", result, "The env variable was not disabled"
+            )
```
Comment (dgibson) [inline selftests/functional/basic.py:1113]: It would also be good to verify that without the override the script does fail, to make sure something's not wrong with the script or its invocation, rather than the environment mangling.

File: avocado/core/__init__.py [hunk 13]
```diff
@@ -178,6 +178,18 @@ def register_core_options():
         help_msg=help_msg,
```
Comment (dgibson) [inline avocado/core/__init__.py:178]: The commit message looks like it needs updating.  AFAICT from the patch itself, it's a different config option rather than a special kwarg to introduce the environment clearing.

File: avocado/plugins/runners/exec_test.py [hunk 14]
```diff
@@ -119,7 +123,11 @@ def _is_uri_a_file_on_cwd(uri):
         return False
 
     def _get_env(self, runnable):
-        env = dict(os.environ)
+        clear_env = runnable.config.get("runner.exectest.clear_env", False)
+        if clear_env:
+            env = {}
+        else:
+            env = dict(os.environ)
```
Comment (dgibson) [inline avocado/plugins/runners/exec_test.py:130]: I don't know enough about the surrounding code to know if this clear is done before or after inserting the Avocado standard environment variables.  i.e. will the final environment have (1) only have variables explicitly set via kwargs, or (2) things explicitly set in kwargs + things that Avocado sets.  I can actually see uses for either case, as well as the original case (3) things set explicitly in kwargs + things set by Avocado + whatever was in the environment to begin with.

File: selftests/functional/basic.py (hunk 12 above)
Comment (richtja) [reply selftests/functional/basic.py:1113]: great idea, I will add that check.

File: avocado/core/__init__.py (hunk 13 above)
Comment (richtja) [reply avocado/core/__init__.py:178]: thank you for catching it. I need to update the commit message to correspond the latest changes.

File: avocado/plugins/runners/exec_test.py (hunk 14 above)
Comment (richtja) [reply avocado/plugins/runners/exec_test.py:130]: Right now, it does the number 2. It will clear environment before adding kwargs and Avocado defaults.

File: selftests/functional/basic.py (hunk 11 above)
Comment (richtja) [reply selftests/functional/basic.py:103]: Yes, that is true. I will update that.

File: avocado/plugins/runners/exec_test.py (hunk 14 above)
Comment (richtja) [reply avocado/plugins/runners/exec_test.py:130]: Sure, we can add (1) as well. I will come up with solution.

**Review by clebergnu — APPROVED** [review]
This version LGTM.  Before merging, I'm asking @dgibson to ACK it too, since he participated a lot in the review and has great interest in the actual result.

**Comment by richtja** [comment]
Hi @clebergnu and @dgibson I have updated this PR, to cover also option (1) from https://github.com/avocado-framework/avocado/pull/5901#discussion_r1569931373. Now the `runner.exectest.clear_env` takes two values `all` and `system`, for clearing the whole environment or only the non-avocado variables. Please have a look, thanks

File: avocado/plugins/runners/exec_test.py [hunk 15]
```diff
@@ -122,6 +122,9 @@ def _get_env(self, runnable):
         env = dict(os.environ)
         if runnable.kwargs:
             env.update(runnable.kwargs)
+            for key, value in runnable.kwargs.items():
+                if value is None:
+                    del env[key]
```
Comment (dgibson) [inline avocado/plugins/runners/exec_test.py:135]: I mentioned this before, but I think it got lost with @clebergnu 's resolved comment.  I believe you need a check for `key in env`, other wise this `del` will generate a `KeyError` if the key wasn't in the original environment.

File: selftests/functional/basic.py [hunk 16]
```diff
@@ -1093,6 +1099,23 @@ def test_non_absolute_path(self):
...
         )
 
+    @skipUnlessPathExists("/bin/bash")
+    def test_env_var_disable(self):
+        with script.TemporaryScript("exec_env_var.sh", EXEC_ENV_VARIABLE_TEST) as tst:
+            res = process.run(
+                f"env TEST_ENV=test avocado-runner-exec-test runnable-run -k exec-test -u {tst}",
+            )
+            result = res.stdout_text.split("\n")[-2]
+            self.assertIn("'returncode': 1", result, "The test might be corrupted.")
```
Comment (dgibson) [inline selftests/functional/basic.py]: It might not be obvious to someone in future what the point of this test is, so I'd be more explicit here.  Maybe `Script unexpectedly passed with environment variable set`.

File: selftests/functional/basic.py [hunk 17]
```diff
@@ -98,6 +98,12 @@ def test(self):
         logging.getLogger("some.other.logger").info("SHOULD BE ON debug.log")
 """
 
+EXEC_ENV_VARIABLE_TEST = """#!/bin/bash
+if [[ -n "${TEST_ENV}"  ]]; then
```
Comment (dgibson) [inline selftests/functional/basic.py]: This of course will arguably-incorrectly pass if `$TEST_ENV` is present but an empty string.  That's hard to fix in shell though, so it's probably acceptable.

File: avocado/plugins/runners/exec_test.py [hunk 18]
```diff
@@ -119,15 +123,19 @@ def _is_uri_a_file_on_cwd(uri):
         return False
 
     def _get_env(self, runnable):
-        env = dict(os.environ)
+        clear_env = runnable.config.get("runner.exectest.clear_env", None)
+        if clear_env in ["all", "system"]:
```
Comment (dgibson) [inline avocado/plugins/runners/exec_test.py:127]: Does anything validate the allowed values for `clear_env`?  If there's no error message, it could be pretty confusing if you made a typo in `system` or `all`.

File: selftests/functional/basic.py [hunk 19]
```diff
@@ -99,9 +99,12 @@ def test(self):
 """
 
 EXEC_ENV_VARIABLE_TEST = """#!/bin/bash
-if [[ -n "${TEST_ENV}"  ]]; then
+if [[ -n "${TEST_ENV}" ]]; then
```
Comment (dgibson) [inline selftests/functional/basic.py:102]: It looks like a spurious space is removed here in the 4th patch, better to fold that fix into patch 3 where the code is introduced.

File: selftests/functional/basic.py [hunk 20]
```diff
@@ -1002,6 +1005,20 @@ def setUp(self):
         )
         self.fail_script.save()
 
+    def _test_env(self, configuration, expected_rc):
+        with script.TemporaryScript("exec_env_var.sh", EXEC_ENV_VARIABLE_TEST) as tst:
+            res = process.run(
+                f"env TEST_ENV=test avocado-runner-exec-test runnable-run -k exec-test -u {tst}",
+            )
+            result = res.stdout_text.split("\n")[-2]
+            self.assertIn("'returncode': 1", result, "The test might be corrupted.")
```
Comment (dgibson) [inline selftests/functional/basic.py:1014]: As in the earlier patch, I'd be more explicit with the error message here.

File: selftests/functional/basic.py [hunk 21]
```diff
@@ -1101,20 +1118,15 @@ def test_non_absolute_path(self):
...
-            self.assertIn(
-                "'returncode': 0", result, "The env variable was not disabled."
-            )
+    @skipUnlessPathExists("/bin/sh")
+    def test_env_clear_system(self):
+        self._test_env('-c \'{"runner.exectest.clear_env": "system"}\'', 0)
+
+    @skipUnlessPathExists("/bin/sh")
+    def test_env_clear_all(self):
+        self._test_env('-c \'{"runner.exectest.clear_env": "all"}\'', 2)
```
Comment (dgibson) [inline selftests/functional/basic.py:1129]: I'm not really seeing how this tests `clear_env = all`.  Don't you need a test that fails in the presence of one of the Avocado default environment variables?

**Review by dgibson — COMMENTED** [review]
In terms of the design and functionality exposed this LGTM.

I still see a few nits in the implementation, for which I've left comments.

File: selftests/functional/basic.py (hunk 17 above)
Comment (richtja) [reply selftests/functional/basic.py]: I agree, but I think that for testing purposes is fine.

File: avocado/plugins/runners/exec_test.py (hunk 18 above)
Comment (richtja) [reply avocado/plugins/runners/exec_test.py:127]: Right, I haven't thought about that, I will add validation, thanks

File: selftests/functional/basic.py (hunk 21 above)
Comment (richtja) [reply selftests/functional/basic.py:1129]: I agree that this test doesn't bring the full coverage. But because of how this feature is implemented, I think that it is enough to check that one of the Avocado default environment variables has been deleted and assume that all of them has been deleted.

File: avocado/plugins/runners/exec_test.py (hunk 15 above)
Comment (richtja) [reply avocado/plugins/runners/exec_test.py:135]: I see your point, but IIUIC the check is not needed, because of line 124. `env.update(runnable.kwargs)`. After this update, the `kwargs` keys become a subset of `env` keys, therefore the key will be always in `env` and we won't get `KeyError`.

**Comment by richtja** [comment]
Hi @dgibson, thank you for your review. I did an update base on your comments. Please have a look.

1. Validation for `runner.exectest.clear_env` values.
2. Updated test message to `Script unexpectedly passed with environment variable set`.
3. Spurious space removal moved to correct commit.

**Review by clebergnu — APPROVED** [review]
LGTM, thanks!

File: selftests/functional/basic.py (hunk 21 above)
Comment (dgibson) [reply selftests/functional/basic.py:1133]: So, I agree that testing one of the default variables is removed is good enough, but I'm not seeing where we're doing that - AFAICT the script only checks for `$TEST_ENV` which is not one of the default ones.  Am I missing something?

**Comment by dgibson** [comment]
Updates LGTM.  Only remaining nit is the question about whether we're really testing the `clear_env = all` case.

**Comment by richtja** [comment]
Hi @dgibson I have split the `EXEC_ENV_VARIABLE_TEST` script into two to make the tests more transparent. Plase have a look.

**Comment by dgibson** [comment]
@richtja well.. I'm trying to ack the latest version, but github appears to be messing up.  I hit "submit review" and the cursor changes, but it never appears to actually go through.

In any case, it LGTM now.  Thanks for the `EXEC_ENV_VARIABLE_TEST` - I did miss the fact that it was already testing that, but I think it's clearer with this change in any case.
